marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/options.xqy

```diff
@@ -564,3 +583,300 @@ declare function merge-impl:propertyspec-to-json($property-spec as element()) as
+                => map:with("path", $path)
+                => map:with("namespaces", fn:head(($merge-rule/namespaces ! xdmp:from-json(.), $namespaces)))
+            )
+          let $implicit-top-level-merges :=
+            for $top-level-property-name in map:keys($property-names-to-values)[fn:not(fn:contains(.,"."))]
+            let $prefix := $top-level-property-name || "."
+            where fn:empty($explicit-merge-rules[map:get(., "propertyName")[. = $top-level-property-name or fn:starts-with(., $prefix)]])
+            return
+              let $to-property-values := $property-names-to-values => map:get($top-level-property-name)
+              let $merge-rule := merge-impl:expand-merge-rule($merge-options, ())
+              let $property-qname := fn:QName(fn:string($target-entity-type-def/namespaceURI), $top-level-property-name)
+              return (
+                xdmp:trace($const:TRACE-MERGE-RESULTS, "Implicit merge for top-level property: " || $top-level-property-name),
+                map:new((
+                  $default-merge-rule-info,
+                  map:entry("propertyName", $top-level-property-name)
+                    => map:with("propertyQName", $property-qname)
+                    => map:with("documentToValuesFunction", $to-property-values)
+                    => map:with("mergeRule", $merge-rule)
+                ))
+              )
+          return ($explicit-merge-rules, $implicit-top-level-merges)
+        )
+  return (
+    if (xdmp:trace-enabled($const:TRACE-MERGE-RESULTS)) then
+      xdmp:trace($const:TRACE-MERGE-RESULTS, "Compiled merge options: " || xdmp:to-json-string($compiled-merge-options))
+    else (),
+    $compiled-merge-options,
+    if (xdmp:trace-enabled($const:TRACE-MERGE-RESULTS)) then
+      xdmp:trace($const:TRACE-MERGE-RESULTS, "Caching compiled merge options with key: " || $cache-id)
+    else (),
+    map:put($_cached-compiled-merge-options, $cache-id, $compiled-merge-options)
+  )
+};
+
+declare function merge-impl:build-prefix-map($source)
+{
+  map:new(
+    for $prefix in ($source ! fn:in-scope-prefixes(.))
+    where fn:not($prefix = "")
+    return
+      map:entry($prefix, fn:namespace-uri-for-prefix($prefix, $source))
+  )
+};
+
+(:
+ : Get a function reference to the default merging function. The function must
+ : be in the http://marklogic.com/smart-mastering/survivorship/merging
+ : namespace.
+ : @param $name  localname of the function to be applied
+ : @param $arity  number of parameters the function takes
+ : @return function reference if found
+ :)
+declare function merge-impl:default-function-lookup(
+  $name as xs:string?,
+  $arity as xs:int
+) as function(*)?
+{
+  fn:function-lookup(
+    fn:QName(
+      "http://marklogic.com/smart-mastering/survivorship/merging",
+      if (fn:exists($name[. ne ""])) then
+        $name
+      else
+        "standard"
+    ),
+    $arity
+  )
+};
+
+declare variable $cached-merging-map as map:map := map:map();
```

Comment:
I think a cache needs a comment describing why it helps. Based on the code below, my assumption would be that build-merging-map is called multiple times, and thus we're avoid a potentially expensive operation - the md5/describe call. But I only see build-merging-map called by compile-merge-options. Is that called enough times to warrant having a cache around?